File: golang/1d_dp/min_cost_climbing_stairs.go
package main

func minCostClimbingStairs(cost []int) int {
	n := len(cost)

	if n == 2 {
		return min(cost[0], cost[1])
	}

	dp1, dp2 := cost[0], cost[1]

	for i := 2; i < n; i++ {
		current := cost[i] + min(dp1, dp2)

		dp1, dp2 = dp2, current
	}

	return min(dp1, dp2)
}

File: java/src/main/java/com/sanedge/interview/dp/MinCostClimbingStairs.java
package com.sanedge.interview.dp;

public class MinCostClimbingStairs {

    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;

        if (n == 2) {
            return Math.min(cost[0], cost[1]);
        }

        int dp1 = cost[0];
        int dp2 = cost[1];

        for (int i = 2; i < n; i++) {
            int current = cost[i] + Math.min(dp1, dp2);

            dp1 = dp2;
            dp2 = current;
        }

        return Math.min(dp1, dp2);
    }
}
